features/ubuntu_pro.feature

```diff
@@ -580,3 +580,84 @@ Feature: Command behaviour when auto-attached in an ubuntu PRO image
            | bionic  | disabled | disabled | disabled | libkrad0  | bundler  | enabled   | cis        |
            | focal   | disabled | n/a      | disabled | hello     | ant      | enabled   | usg        |
            | jammy   | n/a      | n/a      | n/a      | hello     | hello    | enabled   | usg        |
+
+    @series.lts
+    @uses.config.machine_type.gcp.pro
+    @uses.config.machine_type.aws.pro
+    @uses.config.machine_type.azure.pro
+    Scenario Outline: Auto-attach service works on Pro Machine
+        Given a `<release>` machine with ubuntu-advantage-tools installed
+        When I run `systemctl start ua-auto-attach.service` with sudo
+        And I create the file `/etc/ubuntu-advantage/uaclient.conf` with the following:
+        """
+        contract_url: 'https://contracts.canonical.com'
+        data_dir: /var/lib/ubuntu-advantage
+        log_level: debug
+        log_file: /var/log/ubuntu-advantage.log
+        """
+        And I reboot the `<release>` machine
+        And I run `pro status --wait` with sudo
+        And I run `pro security-status --format json` with sudo
+        Then stdout matches regexp:
+        """
+        "attached": true
+        """
+
+        Examples: ubuntu release
+           | release |
+           | xenial  |
+           | bionic  |
+           | focal   |
+           | jammy   |
+
+    @series.lts
+    @uses.config.machine_type.gcp.pro
+    @uses.config.machine_type.aws.pro
+    @uses.config.machine_type.azure.pro
+    Scenario Outline: Auto-attach no-op when cloud-init has ubuntu_advantage on userdata
+        Given a `<release>` machine with ubuntu-advantage-tools installed adding this cloud-init user_data:
+        # This user_data should not do anything, just guarantee that the ua-auto-attach service
+        # # does nothing
+        """
+        ubuntu_advantage:
+        """
+        When I run `cloud-init query userdata` with sudo
+        Then stdout matches regexp:
+        """
+        ubuntu_advantage:
+        """
+        # On GCP, this service will auto-attach the machine automatically after we override
+        # the uaclient.conf file. To guarantee that we are not auto-attaching on reboot
+        # through the ua-auto-attach.service, we are masking it
+        When I run `systemctl mask ubuntu-advantage.service` with sudo
+        When I run `systemctl start ua-auto-attach.service` with sudo
+        And I create the file `/etc/ubuntu-advantage/uaclient.conf` with the following:
+        """
+        contract_url: 'https://contracts.canonical.com'
+        data_dir: /var/lib/ubuntu-advantage
+        log_level: debug
+        log_file: /var/log/ubuntu-advantage.log
+        """
```

Comment:
Shouldn't we need to fix the config file before running the auto-attach service? Right now, `disable_auto_attach` is set when the service is started
```suggestion
        And I create the file `/etc/ubuntu-advantage/uaclient.conf` with the following:
        """
        contract_url: 'https://contracts.canonical.com'
        data_dir: /var/lib/ubuntu-advantage
        log_level: debug
        log_file: /var/log/ubuntu-advantage.log
        """
        When I run `systemctl start ua-auto-attach.service` with sudo
```

Reply:
My idea was to test only the script only during reboot, so when we start the service again the feature flag will block it and only after reboot it should kick off correctly and perform the attach operation

Reply:
Ah okay that makes sense! But then I don't think we need the `When I run `systemctl start ua-auto-attach.service` with sudo` line right? I don't think it should do anything

Reply:
I have added because we stop that service using our custom cloud-init userdata. This is just to guarantee that the reboot will behave as we intend

Reply:
I see and understand now!

I think the `stop` won't affect the next boot though. I'd expect a `disable` or a `mask` to affect next boot, but I'd be surprised if a `stop` did. Did you try the test without that `start` line? (Sorry if you already tested this and I'm just making noise :upside_down_face:)

Reply:
Good point @orndorffgrant, I have added just for safety sake, but you are right, we don't need it.
I have removed that line from the test